rococo-parachains/pallets/xcm-handler/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+pub enum Origin {
+	/// It comes from the relay-chain.
+	RelayChain,
+	/// It comes from a parachain.
+	Parachain(ParaId),
+}
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+
+/// Configuration trait of this pallet.
+pub trait Trait: frame_system::Trait {
+	/// Event type used by the runtime.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// The outer origin type.
+	type Origin: From<Origin>
+		+ From<frame_system::RawOrigin<Self::AccountId>>
+		+ Into<result::Result<Origin, <Self as Trait>::Origin>>;
+
+	/// The outer call dispatch type.
+	type Call: Parameter + Dispatchable<Origin=<Self as Trait>::Origin> + From<Call<Self>>;
+
+	/// The sender of upward messages.
+	type UmpSender: UmpSender;
+
+	/// The sender of horizontal/lateral messages.
+	type HmpSender: HmpSender;
+
+	type Currency: Currency<Self::AccountId>;
+
+	// TODO: Configuration for how pallets map to MultiAssets.
+}
+
+decl_event! {
+	pub enum Event<T> where
+		AccountId = <T as frame_system::Trait>::AccountId
+	{
+		/// Transferred tokens to the account on the relay chain.
+		TransferredToRelayChain(VersionedMultiLocation, VersionedMultiAsset),
+		/// Soem assets have been received.
+		ReceivedAssets(AccountId, VersionedMultiAsset),
+		/// Transferred tokens to the account from the given parachain account.
+		TransferredToParachainViaReserve(ParaId, VersionedMultiLocation, VersionedMultiAsset),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// A version of a data format is unsupported.
+		UnsupportedVersion,
+		/// Asset given was invalid or unsupported.
+		BadAsset,
+		/// Location given was invalid or unsupported.
+		BadLocation,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: <T as frame_system::Trait>::Origin {
+		fn deposit_event() = default;
+
+		/// Transfer some `asset` from the Parachain to the given `dest`.
+		#[weight = 10]
+		fn transfer(origin, dest: VersionedMultiLocation, versioned_asset: VersionedMultiAsset) {
+			let who = ensure_signed(origin)?;
+
+			let asset = MultiAsset::try_from(versioned_asset.clone())
+				.map_err(|_| Error::<T>::UnsupportedVersion)?;
+			let dest = MultiLocation::try_from(dest).map_err(|_| Error::<T>::UnsupportedVersion)?;
+
+			let (asset, amount) = match asset {
+				// The only asset whose reserve we recognise for now is native tokens from the
+				// relay chain, identified as the singular asset of the Relay-chain. From our
+				// context (i.e. a parachain) , this is `Parent`. From the Relay-chain's context,
+				// it is `Null`.
+				MultiAsset::ConcreteFungible { id: MultiLocation::X1(Junction::Parent), amount } =>
+					(MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount }, amount),
+				_ => Err(Error::<T>::BadAsset)?,	// Asset not recognised.
+			};
+
+			let amount: BalanceOf::<T> = amount.checked_into().ok_or(Error::<T>::BadAsset)?;
```

Comment:
Should there be some precision conversion code here?

I Acala, we purposefully stored all the tokens with 18 decimals, including DOT/KSM.
Presumably the unit in XCM messages are the basic unit of the asset, so we want to convert 18 decimals into 10/12 here.